src/item.cpp

```diff
@@ -782,48 +787,81 @@ body_part_set item::get_covered_body_parts() const
 body_part_set item::get_covered_body_parts( const side s ) const
 {
     body_part_set res;
+    iterate_covered_body_parts_internal( s, [&]( const bodypart_str_id & bp ) {
+        res.set( bp );
+    } );
+    return res;
+}
+
+void item::iterate_covered_body_parts_internal( const side s,
+        std::function<void( const bodypart_str_id & )> cb ) const
+{
 
     if( is_gun() ) {
         // Currently only used for guns with the should strap mod, other guns might
         // go on another bodypart.
-        res.set( bodypart_str_id( "torso" ) );
+        cb( bodypart_str_id( "torso" ) );
     }
 
     const islot_armor *armor = find_armor_data();
     if( armor == nullptr ) {
-        return res;
+        return;
     }
+
+    static const std::array<bodypart_str_id, 4> leftSideParts{ {
+            bodypart_str_id{"arm_l"},
+            bodypart_str_id{"hand_l"},
+            bodypart_str_id{"leg_l"},
+            bodypart_str_id{"foot_l"}
+        }
+    };
+    static const std::array<bodypart_str_id, 4> rightSideParts{ {
+            bodypart_str_id{"arm_r"},
+            bodypart_str_id{"hand_r"},
+            bodypart_str_id{"leg_r"},
+            bodypart_str_id{"foot_r"}
+        } };
+
     for( const armor_portion_data &data : armor->data ) {
         if( data.covers.has_value() ) {
-            res.unify_set( data.covers.value() );
+            for( const bodypart_str_id &bpid : data.covers.value() ) {
+                if( !armor->sided ) {
+                    cb( bpid );
+                } else {
+                    switch( s ) {
+                        case side::BOTH:
+                        case side::num_sides:
+                            cb( bpid );
+                            break;
+
+                        case side::LEFT:
+                            [&] {
+                                for( const bodypart_str_id &rightBodyPart : rightSideParts )
+                                {
+                                    if( rightBodyPart == bpid ) {
+                                        return;
+                                    }
+                                }
+                                cb( bpid );
+                            }( );
+                            break;
+
+                        case side::RIGHT:
+                            [&] {
+                                for( const bodypart_str_id &leftBodyPart : leftSideParts )
+                                {
+                                    if( leftBodyPart == bpid ) {
+                                        return;
+                                    }
+                                }
+                                cb( bpid );
```

Comment:
(Optionally) `cb` could return `bool` that can serve as a signal for early termination.

Reply:
I originally thought that would uglify the code a lot but it seems like it'll be pretty clean in this final form it ended in.